assets/js/stripe.js

```diff
@@ -524,11 +524,49 @@ jQuery( function( $ ) {
 					.val( response.source.id )
 			);
 
-			if ( $( 'form#add_payment_method' ).length ) {
-				$( wc_stripe_form.form ).off( 'submit', wc_stripe_form.form.onSubmit );
+			if ( ! $( 'form#add_payment_method' ).length ) {
+				wc_stripe_form.form.submit();
+				return;
 			}
 
-			wc_stripe_form.form.trigger( 'submit' );
+			var apiError = {
+				error: {
+					type: 'api_connection_error'
+				}
+			};
+
+			$.post( {
+				url: wc_stripe_form.getAjaxURL( 'create_setup_intent'),
+				dataType: 'json',
+				data: {
+					stripe_source_id: response.source.id,
+					nonce: wc_stripe_params.add_card_nonce,
+				},
+				error: function() {
+					$( document.body ).trigger( 'stripeError', apiError );
+				}
+			} ).done( function( serverResponse ) {
+				if ( 'success' === serverResponse.status ) {
+					$( wc_stripe_form.form ).off( 'submit', wc_stripe_form.form.onSubmit );
+					return wc_stripe_form.form.submit();
+				} else if ( 'requires_action' !== serverResponse.status ) {
+					return $( document.body ).trigger( 'stripeError', serverResponse );
+				}
+
+				stripe.confirmCardSetup( serverResponse.client_secret, { payment_method: response.source.id } )
+					.then( function( result ) {
+						if ( result.error ) {
+							return $( document.body ).trigger( 'stripeError', result );
+						}
+
+						$( wc_stripe_form.form ).off( 'submit', wc_stripe_form.form.onSubmit );
+						wc_stripe_form.form.trigger( 'submit' );
+					} )
+					.catch( function( err ) {
+						console.log( err );
+						$( document.body ).trigger( 'stripeError', apiError );
```

Comment:
I'm unclear as to why we use `apiError` here instead of `err`. Is it because we just want a generic "Hey we couldn't finish the API request" type error here instead of whatever the api might return?

Reply:
Also, I think we should return the `trigger`, just like in other cases? Or is that not relevant?

Reply:
> Also, I think we should return the trigger, just like in other cases? Or is that not relevant?

Not relevant. All the other `return`s in this Promise chain are there just to stop the excution of the code block, not because the returned value will actually be used.

Reply:
That's right, thank you for the note, @DanReyLop!

> I'm unclear as to why we use apiError here instead of err. Is it because we just want a generic "Hey we couldn't finish the API request" type error here instead of whatever the api might return?

That's mostly because I don't know how to trigger and test that error object, but I see that it's used in `openIntentModal`, so I changed it here as well: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1398/commits/a416a9648e31ebaba22c457bb15ccc8d5f522d14

To be 100% honest though, I still don't know what to expect here :\

Reply:
> Not relevant. All the other `return`s in this Promise chain are there just to stop the excution of the code block, not because the returned value will actually be used.

@DanReyLop @RadoslavGeorgiev How about we put the `return` statements on separate lines then?

I know that we're not actually returning anything, but when the code _looks like_ it's returning something I _expect it_ to return something. I think having the `return` statements on separate lines would improve understanding when skimming the code.

> That's mostly because I don't know how to trigger and test that error object, but I see that it's used in `openIntentModal`, so I changed it here as well: a416a96

That's fair 😅. I'm not sure how to trigger it either! I guess we'll know whenever we see a weird error message 😛.
I do think it'll be more helpful to see the returned error message rather than the generic one, until we know how to test for that failure case at least. Thanks for changing it!

Reply:
> How about we put the return statements on separate lines then?

+1 to that. I've always found that pattern (`return action();` when really you mean `action(); return;`) a bit confusing.

Reply:
Since we've already reviewed this PR pretty thoroughly I decided to change all `return action();` to `action(); return;` where applicable **in `stripe.js` only**, see 9cad6c5907bc2fe7d6146827d459cada87b8d43b.